### PR TITLE
Background API

### PR DESCRIPTION
**Pull request by lbolla**
Various improvements implemented in Molior fork.
See https://github.com/molior-dbs/aptly/tree/molior

* Asynchronous execution of various API tasks to avoid locking the web interface (mirroring, publishing, snapshotting, etc.)
* Batch updates to the database to avoid locking with transactions
* Allow renaming repos
* Improved logging for debugging
* More robust retry strategy

### CODE REVIEW

**Comment by lbolla** [comment]
I ran "system" tests (the ones that used to run on TravisCI) and fixed them to make them pass again (they were broken long time ago).
```
TESTS: 466 SUCCESS: 466 FAIL: 0 SKIP: 37
```
I'm attaching here the full logs for reference.
[system-test-2021-09-22-12-36-29.log.gz](https://github.com/aptly-dev/aptly/files/7210800/system-test-2021-09-22-12-36-29.log.gz)

**Comment by lbolla** [comment]
Just rebased this branch to master.

**Review by neolynx — APPROVED** [review]

**Comment by r4co0n** [comment]
@lbolla, thanks for the additional information about what currently hinders a release, I was wondering. This is relevant for all the folks following #920.

**Comment by r4co0n** [comment]
> If you have time for a documentation PR I would gladly accept it!

I definitely have time for this starting from the weekend of 2021-11-27, before that point I am uncertain, just so you know. :)

File: api/api.go [hunk 1]
```diff
@@ -105,12 +73,95 @@ func acquireDatabase(requests <-chan dbRequest) {
...
+
+func truthy(value interface{}) bool {
+	switch value.(type) {
+	case string:
+		switch strings.ToLower(value.(string)) {
+		case "y", "yes", "t", "true", "1":
+			return true
+		}
+	case int:
+		return value.(int) == 1
```
[the remark this answers is not in the document]
Comment (r4co0n) [reply api/api.go:124]: I would prefer we change this to stay in line with what many folks, coming from C, Python or a pure algebraic logic course, are expecting. ;)

The others are doing it this way should suffice as an argument, your implementation was not concerned about these cases as it wasn't necessary, we can probably save (imagined/assumed) future callers of your function some headache by implementing the same logic as the greybeards.

See e.g. [this](https://www.cs.uic.edu/~jbell/CourseNotes/C_Programming/Decisions.html):
> For interpretation, Zero is interpreted as false and anything non-zero is interpreted as true.

We also case-insensitively interpret the strings `"y", "yes", "t", "true", "1"` as True, somewhat following YAML conventions, or am I mistaken? Please see https://yaml.org/type/bool.html for further values to be considered truthy, if compliance with this standard was aimed for. If not, I strongly recommend we do not follow a part of it misleading people seeing the pattern.

[the remark this answers is not in the document]
Comment (r4co0n) [reply api/api.go:124]: Explictly, we should also match for the string `on`, as it seems.

[the remark this answers is not in the document]
Comment (lbolla) [reply api/api.go]: Thanks for your comments @r4co0n 

I've changed the `truthy` function to be less surprising.

File: api/router.go [hunk 2]
```diff
@@ -81,6 +78,19 @@ func Router(c *ctx.AptlyContext) http.Handler {
...
+		root.GET("/mirrors", apiMirrorsList)
+		root.GET("/mirrors/:name", apiMirrorsShow)
+		root.GET("/mirrors/:name/packages", apiMirrorsPackages)
+		root.POST("/mirrors", apiMirrorsCreate)
+		root.PUT("/mirrors/:name", apiMirrorsUpdate)
+		root.DELETE("/mirrors/:name", apiMirrorsDrop)
+	}
+
+	{
+		root.POST("/gpg/key", apiGPGAddKey)
```
[the remark this answers is not in the document]
Comment (lbolla) [reply api/router.go:91]: Ping @neolynx

**Comment by lbolla** [comment]
@ximon18 As you can see, the CI is not running once I merged your GHA PR. Shall we maybe change the "branch" key in the `on: pull_requests`? Or just have `on: pull_requests` without further specification of the target branch?

**Comment by ximon18** [comment]
> @ximon18 As you can see, the CI is not running once I merged your GHA PR. Shall we maybe change the "branch" key in the `on: pull_requests`? Or just have `on: pull_requests` without further specification of the target branch?

Hi @lbolla, yes I mentioned this [here](https://github.com/aptly-dev/aptly/pull/1010#issuecomment-1021520163).

It depends on the branching policy for development in this repository and when you want the checks to be run:
- If development is sometimes or always done via direct pushes to `main` then `on: pull_requests` will either not be a good match or will not be insufficient.
- If all development is done via pull requests from feature branches to `main`, or if you only care about catching issues just before they hit the `main` branch but not earlier, then you can specify the `main` branch as the target of the `pull_requests` event
- If you use a two level deep branching strategy, e.g. all development is done via pull requests from feature branches into a `dev` branch then you would also need to specify the `dev` branch as a target of the `pull_requests` event.
- If the number and depth of branches used is even deeper and/or varies and you always want changes to be tested then don't specify a target branch for the `pull_requests` event, the workflow will then run for all PRs.

**Comment by ximon18** [comment]
Related to this discussion, on PR #1011 @randombenj just said:
> I would strongly suggest going for on: push as not only main and PRs are tested but every single push so you immediately notice when something breaks ;)

**Pull request merged**